R/permute_cases.R

```diff
@@ -21,38 +21,56 @@ permute_cases.data.frame <- function(cases, n_permutations, feature_distribution
-  }
-  perm <- lapply(seq_along(tokens_lookup), function(i) {
-    remove_n <- sample(doc_size[i], n_permutations, replace = FALSE)
-    remove <- lapply(remove_n, sample, x = doc_size[i] - 1)
-    permutations <- lapply(remove, function(ii) tokens_lookup[[i]][-ii])
-    if (bow) {
-      permutations <- lapply(permutations, function(words) tokens[[i]][!tokens_lower[[i]] %in% words])
-      var_names <- tokens_lookup[[i]]
-    } else {
-      var_names <- paste0(seq_along(tokens_lookup[[i]]), '_', tokens_lookup[[i]])
-    }
-    complete <- rep(1, length(tokens_lookup[[i]]))
-    remove <- do.call(rbind, lapply(remove, function(ii) {complete[ii] <- 0; complete}))
-    remove[1, ] <- 1
-    remove <- Matrix(remove, sparse = TRUE)
-    colnames(remove) <- tokens_lookup[[i]]
-    list(permutations, remove)
-  })
-  removes <- lapply(perm, `[[`, 2)
-  perm <- lapply(perm, `[[`, 1)
-  perm <- stri_join_list(unlist(perm, recursive = FALSE), sep = ' ')
-  perm[seq.int(1, by = n_permutations, length.out = length(cases))] <- cases
-  list(permutations = perm, tabular = removes)
+#' @importFrom purrr map map2 flatten_chr set_names flatten flatten_int map_chr flatten_dbl
+#' @importFrom stringdist seq_dist
+#' @importFrom magrittr %>% set_colnames %<>%
+permute_cases.character <- function(cases, n_permutations, tokenization, bow, dist_fun) {
+  documents_tokens <- map(cases, tokenization) %>%
+{d_tokens <- . ;
+  map2(d_tokens, lengths(d_tokens) %>% cumsum() %>% head(., length(.) - 1) %>% c(0, .),
+       ~ {if (bow) paste0(.x, "_",  seq_along(.) + .y) else unique(.x)})}
+
+  tokens <- documents_tokens %>%
+    flatten_chr() %>%
+    unique(.) %>%
+    sort(decreasing = FALSE) %>%
+    set_names(seq(.))
+
+  documents_tokens %<>%
+    map(~ {d <- . ; which(tokens %in% d) %>%
+      set_names(d, nm = .)})
+
+  dict_size <- length(tokens)
+
+  # Perf on this part should be improved
+  word_selections <- documents_tokens %>%
+    map(~ names(.) %>% as.integer()) %>%
+    map(~ {document <- . ; sample(length(document), as.integer((n_permutations/length(cases)) - 1), replace = T) %>%
+      map(~ sample(document, ., replace = F) %>% sort) %>%
+      c(list(document), .)})
+
+  word_selections_flatten <- flatten(word_selections)
+
+  matrix_cols <- seq(dict_size)
+
+  bow_matrix <- word_selections_flatten %>%
+    map(~ matrix_cols %in% .) %>%
+    flatten_int() %>%
+    Matrix(ncol = dict_size, sparse = TRUE, byrow = TRUE) %>%
+    set_colnames(tokens)
+
+  permutation_candidates <- word_selections_flatten %>% map(~ tokens[.]) %>% map_chr(~ paste(., collapse = " "))
+
+  permutation_distances <- map2(word_selections, documents_tokens,
+                                ~ seq_dist(.x, names(.y) %>% as.integer(), method = dist_fun)) %>%
+    flatten_dbl()
+
+  list(tabular = bow_matrix,
+       permutations = permutation_candidates,
+       word_selections = word_selections,
+       tokens = tokens,
+       permutation_distances = permutation_distances)
 }
+
```

Comment:
So far it s a rewrite in a way I can easily manage for a future rewriting. When I try crazy number of permultations (500K) this part is the slowest by far (40s on my i7), I want to remake it using Rcpp, that's why I divided the code that way.